qiskit/optimization/converters/integer_to_binary.py

```diff
@@ -215,14 +215,12 @@ def decode(self, result: OptimizationResult) -> OptimizationResult:
         Returns:
             The result of the original problem.
         """
-        vals = result.x
-        new_vals = self._decode_var(vals)
-        result.x = new_vals  # type: ignore
+        result._val = self._decode_var(result.x)  # type: ignore
         return result
```

Comment:
Any chance we can avoid accessing `_val` directly?

Reply:
It is difficult because this method receives an object of various `OptimizationResult` children such as ADMMOptimizationResult, etc. and interprets the values of binary variables into integer variables.
So, it is hard to use the constructor of the children. But, I can think of a helper function to do it. Let me think...

Reply:
If we make a following copy method, we can avoid accessing `_val`. It may be better than a public setter of `self.x`. What do you think? 
```python
class OptimizationResult:
...
    def copy(self, x: Union[List[float], np.ndarray]) -> OptimizationResult:
        new_result = copy.deepcopy(self)
        new_result._val = x
        return new_result
```

Reply:
If no algorithm extends `OptimizationResult`, it is easy. We just call the constructor with updated values.

Reply:
This is a very good question. Do we need sub-classes of `OptimizationResult`? Frankly I don't really get the benefits of sub-classing results.
Anyway, if we can't come up with a better solution we can keep direct access, it is just python, everything is public no matter how hard we try.

Reply:
Oh, I'd say adding `copy` we make it more complicated. I prefer to have direct access :)

Reply:
> 
> 
> If no algorithm extends `OptimizationResult`, it is easy. We just call the constructor with updated values.

That's the question. What's the reason for sub-classes? I don't see serious reasons for this.

Reply:
It's because different algorithms output `OptimizationResults` with  different additional informations. For example, `ADMMOptimizationResults` has `state` property, and `MinimumEigenOptimizerResult` has `samples` property etc. We can create an `OptimizationResults` class with lots of properties that most of those property will be empty most of the times. I'm not sure which way is better, crating sub-classes or creating an `OptimizationResult` class with lots of properties.

Reply:
@a-matsuo Matsuo-san, In case of ADMM we have a sub-class, but we fill in both `results` and `state` with the absolutely same value. So, in case of ADMM I don't see why we have this, except that Stefan told me to do so :)
I don't think we need many-many properties defined in the `OptimizationResult` class. I believe we only need `results` as a generic property where we may keep any value we want. And as a conclusion I don't see a reason why we need sub-classes, I think we may have just a basic `OptimizationResult` for all purposes.

Reply:
In Aqua we used to have a dictionary return for the algorithms which we are trying to go away from. This is all very well and extensible but hard to know what the keys are and the IDE can't help you the same way it can do with getters which are much easy to see/discover. Hence we went over to result classes. The intent was to keep info the user would be interested in visible this way. And so for example we have a commonn min eigen result which VQE extends and adds unique fields it has. Same for QPE etc. The more it goes to some generic catch all the less visible things can become. Now in chemistry result there is an algorithm result field that contains whatever the raw result was of Aqua algorithm used - maybe this akin to the results field you are talking about.

Reply:
So you have a long story of using result classes in Aqua. In the optimization stack we have a similar issue, the difference to chemistry, as I got it, that we have two mandatory(if everything goes well) properties, a float and an array of floats(`fval` is a cost function value and `x` is a solution vector). The third parameter(`results`) is up to an optimizer. It can be everything. Most of time, I guess, those two properties are the things a user may need(may be I'm wrong here). That's why I suggest to keep `fval` and `x` as mandatory properties with certain type hints and have `results` as an optional and unspecified. If somebody wants and knows the concrete result class it is still possible to add a type hint to resolve `results` type. It is not a great solution, not cast in stone but python is a serious OO language. 
In overall, if you think we should keep as many type hints as possible at a cost of some overhead - fine with me, I like to know what I have as an output.

Reply:
In my mind it would be say whether an SLSPQOptimzer or ADMMOptimizer would want/need to subclass the generic optimizer result to add unique fields that may be of interest/use to a user when you know its the specific optimizer you used, We did this via subclassing in Aqua core algorithms so VQE has a result, class and so does QPE etc, each extending a more generic result. For chemistry application module we just put the 'raw' algorithm result - ie base class from which all Aqua core algos derive https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/chemistry/core/chemistry_operator.py#L111
For me its less about typehints and stuff, but more about the user being able to find/discover result parts which can be more easily seen through some public API with setters and is more opaque with a dictionary. What the right balance is for Optimization in making say specific getters versus dropping a dictionary or other complex result object into some generic results field I cannot say. I am merely noting some of the thinking around what we did so far for Aqua (and still are in the process of conversion to in some cases)

Reply:
I see, thanks for sharing this. In the optimization stack we don't have a dictionary. But anyway, since you've already implemented a sub-classing pattern, let's keep it in optimization also.

Reply:
If we keep subclass pattern, we need to choose one of the following options.
1. Allow a direct access `OptimizationResult._val` to only `IntegerToBinaryConverter.decode` as an exceptional case.
2. Make a public setter of `OptimizationResult.x`
3. Make a copy method as I wrote https://github.com/Qiskit/qiskit-aqua/pull/1131#discussion_r458893385

I think `OptimizationRestuls._results` is regarded as the raw results of each algorithm. So, how about changing the name of the class variable as `OptimizationResults._raw_results` and the getter as `OptimizationResults.raw_results` to be more clear? If we prohibit sub-classes of OptimizationResult and make algorithms use `_raw_data` to embed algorithm specific data or `AlgorithmResult` as @woodsp-ibm mentioned, we can easily interpret the data without direct access of `_val`. https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/aqua/algorithms/algorithm_result.py#L25

Reply:
I'm in favor of 1. in this case.
Let's keep sub-classing.

Reply:
I notice a technique to avoid the direct access!
For example, ADMM currently generates a result as follows.
```python
result = ADMMOptimizationResult(solution, objective_value, self._state)
# convert back integer to binary
result = cast(ADMMOptimizationResult, int2bin.decode(result)) # `int2bin.decode` needs to access `_val` because `result is ADMMOptimizationResult.
```
https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/optimization/algorithms/admm_optimizer.py#L361

By changing the logic, `int2bin` interprets the base class `OptimizationResult` by calling the constructor. We can also delete `cast` too.
```
# convert back integer to binary
base_result = OptimizationResult(solution, objective_value)
base_result = int2bin.decode(base_result)  # `int2bin.decode` can make a new result by calling the constructor of OptimizationResult
result = ADMMOptimizationResult(base_result.x, base_result.fval, self._state)
```

It works for MinimumEigenOptimizer too. I made a branch to try this technique.
commit: https://github.com/t-imamichi/qiskit-aqua/commit/2ae7f5faf8e921f9e41bdcc7fa0834193dc8f171
branch: https://github.com/t-imamichi/qiskit-aqua/tree/opt_result2

What do you think?

Reply:
If you are OK. I merge the branch to this pullreq to continue the discussion.

Reply:
@t-imamichi maybe you can also merge #1122 to this branch to resolve the conflicts,
or if you want, I can merge this branch to #1122

Reply:
Sorry, I missed this comment. Looks fine to me.

Reply:
I adjust the code with the merged PR #1122